### PR TITLE
bpo-1635741: Fix potential refleaks in binascii module

### PR DESCRIPTION
**Pull request by shihai1991**
https://bugs.python.org/issue1635741

### CODE REVIEW

File: Modules/binascii.c [hunk 1]
```diff
@@ -1639,16 +1649,46 @@ static PyModuleDef_Slot binascii_slots[] = {
...
+    Py_VISIT(state->Error);
+    Py_VISIT(state->Incomplete);
+    return 0;
+}
+
+static int
+binascii_clear(PyObject *m)
+{
+    binascii_state *state = get_binascii_state(m);
+    if (state == NULL) {
```
Comment (vstinner) [inline Modules/binascii.c]: I don't think that state can be NULL. If it's NULL, it's a bug in Python. No? Same remark in binascii_traverse().

File: Modules/binascii.c [hunk 2]
```diff
@@ -1639,16 +1649,46 @@ static PyModuleDef_Slot binascii_slots[] = {
     {0, NULL}
 };
 
+static int
+binascii_traverse(PyObject *m, visitproc visit, void *arg)
```
Comment (vstinner) [inline Modules/binascii.c]: nitpick: I would prefer to use longer variable name than just "m". Rename "m" to "mod" or "module". Same remark if following functions.

Comment (shihai1991) [reply Modules/binascii.c]: ok, no probleam, i will update it.

File: Modules/binascii.c [hunk 3]
```diff
@@ -1639,16 +1649,46 @@ static PyModuleDef_Slot binascii_slots[] = {
...
+{
+    binascii_state *state = get_binascii_state(m);
+    if (state == NULL) {
```
Comment (shihai1991) [inline Modules/binascii.c]: I will update this behavior soon.

File: Modules/binascii.c (hunk 1 above)
Comment (shihai1991) [reply Modules/binascii.c]: I do some debug in my vm:
```
Breakpoint 1, binascii_traverse (module=0x7ffff7e803d0, visit=0x4630b6 <bad_traverse_test>, arg=0x0) at /temp/shihai/cpython/Modules/binascii.c:1654
1654    {
(gdb) n
1655        binascii_state *state = get_binascii_state(module);
(gdb) n
1656        if (state == NULL) {
(gdb) n
1657            return -1;
(gdb) p state
$5 = (binascii_state *) 0x0
(gdb) c
Continuing.

Breakpoint 3, binascii_exec (module=0x7ffff7e803d0) at /temp/shihai/cpython/Modules/binascii.c:1615
1615    binascii_exec(PyObject *module) {
```

What's the reason? 
the malloc operation of `md_state` will be called after first calling `binascii_traverse`.
https://github.com/python/cpython/blob/master/Objects/moduleobject.c#L399.

**Comment by vstinner** [comment]
I created https://bugs.python.org/issue39824 to change the md_state==NULL case.

**Pull request merged**